### DISCUSSION TITLE
lightsleep and 2 processors

### DISCUSSION THREAD

**Discussion by CrabbyPete** [micropython/micropython#14020] [RP2040 / Pico]
I am using both processors on an RP Pico.  One reads just reads from a UART and the other manages everything. If I use machine.lightsleep in one process does it cause both processors to sleep or only one?

**Comment by robert-hh** [comment]
The implementation of lightsleep() affects the clocks. It stops some clocks and slows down others. So yes, lightsleep affects both cores.

**Comment by mungewell** [comment]
For reference; this is code that microPython uses to enter `lightsleep()`
https://github.com/micropython/micropython/blob/master/ports/rp2/modmachine.c#L108

Basically it converts each clock to use the 'raw' XTAL frequency, shuts down the PLLs, and then sets **only** `CLOCKS_SLEEP_EN0_CLK_RTC_RTC_BITS` and `CLOCKS_SLEEP_EN1_CLK_SYS_TIMER_BITS` to be active during sleep (so it can wake in X ms specified).


I started looking at this, as like you my code runs on both cores. But I also do a LOT of work in the PIO blocks, which have their own clocks (slaved from SYS_CLK). For me there are significant chunks of time where my CPU code could be in light sleep, and the PIO left running...
![PIO_clocks_in_lightsleep](https://github.com/micropython/micropython/assets/236907/8d37d2c0-6d99-4b9b-9ec6-bc616266c2d1)


I can see how 'we' might make some changes to this code and plan to investigate when this really saves me any power (as running from battery)

**Reply by mungewell** [reply]
Did some poking around in the microPython source and I think that this idea has wings...
https://github.com/micropython/micropython/issues/14079#issuecomment-1993254298

Test with blinking LED, powered from VSYS at 3.8V:
Before patch, PIO stopped: 20mA, then 3mA in lightsleep() and 23mA after (faster clock speed?).
After patch, PIO runs though out: 20mA, then 8mA in lightsleep() and then 20mA after.

So if it (truely) does not break my code, then I would be saving ~12mA consumption!
